src/psyclone/tests/psyir/frontend/fparser2_test.py

```diff
@@ -2752,6 +2826,147 @@ def test_structures(fortran_reader, fortran_writer):
         "  PROCEDURE, NOPASS :: test_code\n"
         "END TYPE test_type\n" in result)
 
+    # public type that contains procedures
+    test_code = (
+        "module test_mod\n"
+        "  use kernel_mod, only : parent_type\n"
+        "  type, extends(parent_type) :: test_type\n"
+        "    integer :: i = 1\n"
+        "    contains\n"
+        "      procedure :: test_code\n"
+        "      procedure, private :: test_code_private\n"
+        "      procedure, public :: test_code_public\n"
+        "  end type test_type\n"
+        "  contains\n"
+        "  subroutine test_code()\n"
+        "  end subroutine\n"
+        "  subroutine test_code_private()\n"
+        "  end subroutine\n"
+        "  subroutine test_code_public()\n"
+        "  end subroutine\n"
+        "end module test_mod\n")
+    psyir = fortran_reader.psyir_from_source(test_code)
+    sym_table = psyir.children[0].symbol_table
+    symbol = sym_table.lookup("test_type")
+    assert isinstance(symbol, DataTypeSymbol)
+    assert isinstance(symbol.datatype, StructureType)
+    assert len(symbol.datatype.procedure_components) == 3
+    result = fortran_writer(psyir)
+    assert (
+        "  type, extends(parent_type), public :: test_type\n"
+        "    integer, public :: i = 1\n"
+        "    contains\n"
+        "      procedure, public :: test_code\n"
+        "      procedure, private :: test_code_private\n"
+        "      procedure, public :: test_code_public\n"
+        "  end type test_type\n" in result)
+
+    # private type that contains procedures
+    # the `test_code` procedure should thus become private
```

Comment:
I don't think this is correct. You've put the `private` attribute on the declaration of `test_type` so `test_type` will be private - this says nothing about the visibility of the components within it. What I meant was:
```fortran
type my_type
  private
  contains
     procedure :: this_should_be_private
end type
```

Reply:
Oops, I had misunderstood. Indeed, components should not inherit the derived type visibility and my implementation was plain wrong. Thanks for spotting this!

I checked the standard and implemented visibility for procedure components using a "standalone" `private` in the contains part.
However, note before reviewing that `private` statements in the data components and procedure components parts are independent (see e.g. NOTE 4.44 p.74 of https://dci.dci-gitlab.cines.fr/webextranet/_downloads/91b30eb00b39f6896fae528e5d0d66fc/Fortran_2003_N1601.pdf) so that your code above would not modify the visibility of `this_should_be_private` whereas this would:
```fortran
type my_type
  contains
     private
     procedure :: this_should_be_private
end type
```

I've fixed this and added tests where the standalone `private` statement either does or does not impact the visibility of a procedure and of a data component.